ChronoVisor/src/ChronicleMetaDirectory.cpp

```diff
@@ -40,65 +37,98 @@ bool ChronicleMetaDirectory::create_chronicle(const std::string& name) {
     t2 = std::chrono::steady_clock::now();
     std::chrono::duration<double, std::nano> duration = (t2 - t1);
     LOGD("time in %s: %lf ns", __FUNCTION__, duration.count());
-    return res.second;
+    if (res.second) {
+        return CL_SUCCESS;
+    } else {
+        return CL_ERR_UNKNOWN;
+    }
 }
 
-bool ChronicleMetaDirectory::create_chronicle(const std::string& name,
-                                              const std::unordered_map<std::string, std::string>& attrs) {
+int ChronicleMetaDirectory::create_chronicle(const std::string& name,
+                                             const std::unordered_map<std::string, std::string>& attrs) {
     LOGD("creating Chronicle name=%s", name.c_str());
     Chronicle *pChronicle = new Chronicle();
     pChronicle->setName(name);
     pChronicle->setProperty(attrs);
     uint64_t cid = CityHash64(name.c_str(), name.size());
     pChronicle->setCid(cid);
     auto res = chronicleMap_->emplace(cid, pChronicle);
-    if (chronicleMap_->size() % 10 == 0)
-        LOGD("10 chronicles have been created");
-    return res.second;
+    if (res.second) {
+        return CL_SUCCESS;
+    } else {
+        return CL_ERR_UNKNOWN;
+    }
 }
 
-bool ChronicleMetaDirectory::destroy_chronicle(const std::string& name,
-                                               int flags) {
+int ChronicleMetaDirectory::destroy_chronicle(const std::string& name,
+                                              int& flags) {
     LOGD("destroying Chronicle name=%s", name.c_str());
     std::chrono::steady_clock::time_point t1, t2;
     t1 = std::chrono::steady_clock::now();
     uint64_t cid = CityHash64(name.c_str(), name.size());
     auto chronicleRecord = chronicleMap_->find(cid);
     if (chronicleRecord != chronicleMap_->end()) {
         Chronicle *pChronicle = chronicleRecord->second;
+        if (pChronicle->getAcquisitionCount() != 0) {
+            return CL_ERR_ACQUIRED;
+        }
         delete pChronicle;
```

Comment:
Potential race condition here : what if you delete the chronicle in memory and another client tries to create a chronicle with the same cid (name)? Can you update the chronicle class in memory and the map entry atomically?